src/ir/principal-type.h

```diff
@@ -0,0 +1,134 @@
+
+// A heap type that is a subtype of all other heap types, no matter what
+// hierarchy they belong to.
+struct BottomHeapType : std::monostate {};
+
+// Either a nullability or a nullability type variable.
+using VarNullability = std::variant<Nullability, Index>;
+
+// Bottom sharedness, a normal sharedness, or a sharedness type variable.
+using VarSharedness = std::variant<BottomShare, Shareability, Index>;
+
+// An abstract heap type with some given sharedness.
+struct VarAbsHeapType {
+  VarSharedness share;
+  HeapType ht;
+
+  bool operator==(const VarAbsHeapType& other) const {
+    return share == other.share && ht == other.ht;
+  }
+  bool operator!=(const VarAbsHeapType& other) const {
+    return !(*this == other);
+  }
+};
+
+// Either an exactness or an exactness type variable.
+using VarExactness = std::variant<Exactness, Index>;
+
+// A defined heap type with a given exactness.
+struct VarDefHeapType {
+  VarExactness exact;
+  HeapType ht;
+
+  bool operator==(const VarDefHeapType& other) const {
+    return exact == other.exact && ht == other.ht;
+  }
+  bool operator!=(const VarDefHeapType& other) const {
+    return !(*this == other);
+  }
+};
+
+// A bottom heap type, an abstract heap type, a defined heap type, or a heap
+// type variable.
+using VarHeapType =
+  std::variant<BottomHeapType, VarAbsHeapType, VarDefHeapType, Index>;
+
+// A reference comprising a given nullability and heap type.
+struct VarRef {
+  VarNullability null;
+  VarHeapType ht;
+
+  bool operator==(const VarRef& other) const {
+    return null == other.null && ht == other.ht;
+  }
+  bool operator!=(const VarRef& other) const { return !(*this == other); }
+};
+
+// A normal type, a reference type using the extended syntax, or a type
+// variable. If a reference type can be expressed in the normal syntax rather
+// than the extended syntax, it should be represented here as a `Type`, not a
+// `VarRef`.
+using VarType = std::variant<Type, VarRef, Index>;
+
+// A principal type, including extended input and output types for an
+// instruction sequence and whether the sequence is stack-polymorphic, i.e.
+// contains an instruction that unconditionally transfers control flow out of
+// the sequence, i.e. is unreachable. The parameters are stored in reverse order
+// for efficiency. Variable indices for each kind of variable must be introduced
+// contiguously in order starting at zero from the end to beginning of the
+// params (beginning to end of the reversed params).
+struct PrincipalType {
+  std::vector<VarType> rparams;
+  std::vector<VarType> results;
+  bool unreachable;
+
+  PrincipalType(std::initializer_list<VarType> params,
+                std::initializer_list<VarType> results,
+                bool unreachable = false)
+    : rparams(params), results(results), unreachable(unreachable) {
+    std::reverse(rparams.begin(), rparams.end());
+  }
+
+  // Update this type to be the composition of this and `next`.
```

Comment:
What is a "composition" of two principal types? `<X, Y>` as a tuple, or something more?

Reply:
I see, thanks, so this is composition in the sense of the composing the instructions, not the types.
```suggestion
  // Update this type to be the composition of this and `next` (that is, the
  // principal type of the composed two corresponding sequences of
  // instructions).
```

Reply:
There are no instructions, so it really is composing the types. I think the meaning of composition of function types should be fairly clear, though. For example the composition of `t1 -> t2` and `t2 -> t3` is `t1 -> t3`. Is that not what you would expect?

Reply:
Fair enough, as I said my first intuition here was that this meant composition as a tuple, but maybe that's just me...